README.md
# @mapbox/fontmachine

[![Build Status](https://travis-ci.org/mapbox/fontmachine.svg?branch=master)](https://travis-ci.org/mapbox/fontmachine) [![codecov](https://codecov.io/gh/mapbox/fontmachine/branch/master/graph/badge.svg)](https://codecov.io/gh/mapbox/fontmachine)

Make GL-ready pbfs and metadata for usage in fontstack API.

### `makeGlyphs(opts, callback)`

Make all metadata (codepoints) and SDF PBFs necessary for Mapbox GL fontstacks.

### Parameters

| parameter            | type                     | description                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
| -------------------- | ------------------------ | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `opts`               | `Object`                   | Options object with required `font` and `filetype` properties, and any optional parameters.                                                                                                                                                                                                                                                                                                                                                                                         |
| `opts.font`          | `Buffer`                   | The font file as a `Buffer`.                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| `opts.filetype`      | `string`                   | Type of font (e.g. `'.ttf'` or `'.otf'`).                                                                                                                                                                                                                                                                                                                                                                                                                                           |
| `[opts.concurrency]` | `number`                   | _optional:_ Concurrency to use when processing font into PBFs. If `undefined`, concurrency is unbounded.                                                                                                                                                                                                                                                                                                                                                                            |
| `callback`           | `function(err, result)` | Callback takes arguments (`error`, `result`).

### Callback `result`

| property   | type                                  | description                                                                                 |
| ---------- | ------------------------------------- | ------------------------------------------------------------------------------------------- |
| `name`     | `string`                              | The name of this font (concatenated family_name + style_name).                              |
| `stack`    | `Array<{name: string, data: Buffer}>` | An array of 256 `{name: filename, data: buffer}` objects with SDF PBFs covering points 0-65535. |
| `metadata` | `{family_name: string, style_name: string}`                              | Metadata about the font.                                  |
| `codepoints` | `Array<int>` | Array of codepoints covered by the font.
| `original` | `{name: string, data: Buffer}`        | An object containing the original font file (named `'original<filetype>'`)                 |

## Installation

Requires [nodejs](http://nodejs.org/).

```sh
$ npm install @mapbox/fontmachine
```

## Tests

```sh
$ npm test
```




CHANGELOG.md
# Changelog

## v0.3.0

- Add support for Node 10. Drop support for Node 4 and Node 6.
- Add `concurrency` option to the `makeGlyphs` function, to allow specifying the number of glyph ranges that will be processed in parallel.
- Perform some code and documentation maintenance.

## v0.2.2

- Upgraded node-fontnik from 0.5.0 to 0.5.2
- Adds package-lock.json

## v0.2.1

- Now using @mapbox/fontmachine namespace, `fontmachine` is deprecated

## v0.2.0

- Upgraded node-fontnik from ^0.4.3 to ~0.5.0
- Now only testing on node v4/v5


test/test.js
'use strict';

const fs = require('fs');
const path = require('path');
const zlib = require('zlib');

const sinon = require('sinon');
const test = require('tap').test;
const fontnik = require('fontnik');
const rewire = require('rewire');
const fontmachine = rewire('../index.js');

const glyphComposite = require('@mapbox/glyph-pbf-composite');
const opensans = fs.readFileSync(path.resolve(path.join(__dirname, '/fixtures/fonts/open-sans/OpenSans-Regular.ttf')));
const guardianbold = fs.readFileSync(path.resolve(path.join(__dirname, '/fixtures/fonts/GuardianTextSansWeb/GuardianTextSansWeb-Bold.ttf')));
const noto = fs.readFileSync(path.resolve(path.join(__dirname, '/fixtures/fonts/noto/NotoSans-Regular.ttc')));
const invalid = fs.readFileSync(path.resolve(path.join(__dirname, '/fixtures/fonts/invalid/foo.ttf')));

test('handle undefined style_name', (t) => {
  fontmachine.makeGlyphs({ font: guardianbold, filetype: '.ttf' }, (err, font) => {
    t.ifError(err);
    t.ok(font, 'returns a font');
    t.equal(font.name, '?');
    t.ok(font.metadata, 'metadata exists');
    t.ok(font.metadata instanceof Object, 'metadata is object');
    t.equal(font.metadata.family_name, '?');
    t.equal(font.metadata.style_name, undefined);
    t.ok(font.codepoints, 'codepoints exists');
    t.ok(Array.isArray(font.codepoints), 'codepoints is array');
    t.equal(font.codepoints.length, 227, 'right number of codepoints');
    t.ok(font.original, 'original font file exists');
    t.equal(font.original.name, 'original.ttf');
    t.deepEqual(font.original.data, guardianbold, 'Font files equal eachother');
    t.ok(font.stack, 'pbf stack exists');
    t.equal(font.stack.length, 256, 'font stack has the right size');
    t.ok(Array.isArray(font.stack), 'pbf stack is array');
    t.end();
  });
});

test('handle api misuse opts.font', (t) => {
  t.throws(() => {
    fontmachine.makeGlyphs({ font: 'STRING' });
  }, /opts.font must be a Buffer/);
  t.end();
});

test('handle api misuse opts.filetype', (t) => {
  t.throws(() => {
    fontmachine.makeGlyphs({ font: Buffer.alloc(0), filetype: null });
  }, /opts.filetype must be a String/);
  t.end();
});

test('handle api misuse callback', (t) => {
  t.throws(() => {
    fontmachine.makeGlyphs({ font: Buffer.alloc(0), filetype: '.ttf' }, 'callback');
  }, /Callback must be a Function/);
  t.end();
});

test('handle not a font file', (t) => {
  fontmachine.makeGlyphs({ font: invalid, filetype: '.ttf' }, (err) => {
    t.ok(err);
    t.equal(err.message, 'could not open font file', 'Handles file that is not a font');
    t.end();
  });
});

test('handle multifont', (t) => {
  fontmachine.makeGlyphs({ font: noto, filetype: '.ttc' }, (err) => {
    t.ok(err);
    t.equal(err.message, 'Multiple faces in a font are not yet supported.', 'Calls callback with an error');
    t.end();
  });
});

test('calls-back with fontnik.range error', (t) => {
  const mockError = new Error();
  sinon.stub(fontnik, 'range').callsFake((_, cb) => cb(mockError));

  fontmachine.makeGlyphs({ font: opensans, filetype: '.ttf' }, (err, font) => {
    t.ok(err, 'has an error');
    t.equal(err, mockError, 'error passed along from fontnik.range');
    t.notOk(font, 'does not have font result');
    sinon.restore();
    t.done();
  });
});

test('calls-back with zlib.gzip error', (t) => {
  const mockError = new Error();
  sinon.stub(zlib, 'gzip').callsFake((_, cb) => cb(mockError));

  fontmachine.makeGlyphs({ font: opensans, filetype: '.ttf' }, (err, font) => {
    t.ok(err, 'has an error');
    t.equal(err, mockError, 'error passed along from zlib.gzip');
    t.notOk(font, 'does not have font result');
    sinon.restore();
    t.done();
  });
});

test('font machine 256-511', (t) => {
  fontmachine.makeGlyphs({ font: opensans, filetype: '.ttf' }, (err, font) => {
    t.ifError(err);
    t.ok(font, 'returns a font');
    t.equal(font.name, 'Open Sans Regular');
    t.ok(font.metadata, 'metadata exists');
    t.ok(font.metadata instanceof Object, 'metadata is object');
    t.equal(font.metadata.family_name, 'Open Sans');
    t.equal(font.metadata.style_name, 'Regular');
    t.ok(font.codepoints, 'codepoints exists');
    t.ok(Array.isArray(font.codepoints), 'codepoints is array');
    t.equal(font.codepoints.length, 882, 'right number of codepoints');
    t.ok(font.original, 'original font file exists');
    t.equal(font.original.name, 'original.ttf');
    t.deepEqual(font.original.data, opensans, 'Font files equal eachother');
    t.ok(font.stack, 'pbf stack exists');
    t.equal(font.stack.length, 256, 'font stack has the right size');
    t.ok(Array.isArray(font.stack), 'pbf stack is array');
    const stack512 = font.stack.filter((stack) => {
      return stack.name === '256-511.pbf';
    })[0];
    zlib.gunzip(stack512.data, (err, rawBody) => {
      t.ifError(err, 'gunzips successfully');
      const decodedStack = glyphComposite.decode(rawBody);
      t.ok(decodedStack.stacks, 'stack exists');
      t.equal(decodedStack.stacks.length, 1, 'Only one stack');
      t.ok(decodedStack.stacks[0].glyphs, 'glyphs exist');
      t.equal(decodedStack.stacks[0].glyphs.length, 140, 'There are glyphs');
      t.equal(decodedStack.stacks[0].name, 'Open Sans Regular');
      t.equal(decodedStack.stacks[0].range, '256-511');
      t.equal(decodedStack.stacks[0].glyphs[0].id, 256, 'correct first id');
      t.equal(decodedStack.stacks[0].glyphs[0].width, 15, 'correct first width');
      t.equal(decodedStack.stacks[0].glyphs[0].height, 20, 'correct first height');
      t.equal(decodedStack.stacks[0].glyphs[0].left, 0, 'correct first left');
      t.equal(decodedStack.stacks[0].glyphs[0].top, -6, 'correct first top');
      t.equal(decodedStack.stacks[0].glyphs[0].advance, 15, 'correct first advance');
      t.equal(decodedStack.stacks[0].glyphs[75].id, 331, 'correct 76th id');
      t.equal(decodedStack.stacks[0].glyphs[75].width, 11, 'correct 76th width');
      t.equal(decodedStack.stacks[0].glyphs[75].height, 19, 'correct 76th height');
      t.equal(decodedStack.stacks[0].glyphs[75].left, 2, 'correct 76th left');
      t.equal(decodedStack.stacks[0].glyphs[75].top, -13, 'correct 76th top');
      t.equal(decodedStack.stacks[0].glyphs[75].advance, 14, 'correct 76th advance');
      t.end();
    });
  });
});

test('font machine 7680-7935', (t) => {
  fontmachine.makeGlyphs({ font: opensans, filetype: '.ttf' }, (err, font) => {
    t.ifError(err);
    const stack512 = font.stack.filter((stack) => {
      return stack.name === '7680-7935.pbf';
    })[0];
    zlib.gunzip(stack512.data, (err, rawBody) => {
      t.ifError(err, 'gunzips successfully');
      const decodedStack = glyphComposite.decode(rawBody);
      t.ok(decodedStack.stacks, 'stack exists');
      t.equal(decodedStack.stacks.length, 1, 'Only one stack');
      t.ok(decodedStack.stacks[0].glyphs, 'glyphs exist');
      t.equal(decodedStack.stacks[0].glyphs.length, 100, 'There are glyphs');
      t.equal(decodedStack.stacks[0].name, 'Open Sans Regular');
      t.equal(decodedStack.stacks[0].range, '7680-7935');
      t.equal(decodedStack.stacks[0].glyphs[0].id, 7680, 'correct first id');
      t.equal(decodedStack.stacks[0].glyphs[0].width, 15, 'correct first width');
      t.equal(decodedStack.stacks[0].glyphs[0].height, 24, 'correct first height');
      t.equal(decodedStack.stacks[0].glyphs[0].left, 0, 'correct first left');
      t.equal(decodedStack.stacks[0].glyphs[0].top, -9, 'correct first top');
      t.equal(decodedStack.stacks[0].glyphs[0].advance, 15, 'correct first advance');
      t.equal(decodedStack.stacks[0].glyphs[75].id, 7905, 'correct 76th id');
      t.equal(decodedStack.stacks[0].glyphs[75].width, 14, 'correct 76th width');
      t.equal(decodedStack.stacks[0].glyphs[75].height, 18, 'correct 76th height');
      t.equal(decodedStack.stacks[0].glyphs[75].left, 1, 'correct 76th left');
      t.equal(decodedStack.stacks[0].glyphs[75].top, -8, 'correct 76th top');
      t.equal(decodedStack.stacks[0].glyphs[75].advance, 14, 'correct 76th advance');
      t.end();
    });
  });
});

test('can set processing concurrency', (t) => {
  const opts = {
    font: opensans,
    filetype: '.ttf',
    concurrency: 1 // process glyph ranges serially
  };

  const writeGlyphs = fontmachine.__get__('writeGlyphs');
  const writeGlyphsSpy = sinon.spy(writeGlyphs);
  const unset = fontmachine.__set__('writeGlyphs', writeGlyphsSpy);

  fontmachine.makeGlyphs(opts, (err, font) => {
    t.ifError(err);
    t.ok(font, 'returns a font');

    t.ok(writeGlyphsSpy.called, 'writeGlyphs is called');

    const calls = writeGlyphsSpy.getCalls();
    t.equal(calls.length, 256, 'writeGlyphs called expected number of times');

    for (let i = 0; i < calls.length; i++) {
      // ensure that all calls to writeSpy were processed in sequential order
      const call = calls[i];
      t.equal(call.args[1], i * 256, 'start argument');
      t.equal(call.args[2], (i + 1) * 256 - 1, 'end argument');
    }

    unset();
    sinon.restore();
    t.done();
  });
});


index.js
'use strict';

const fontnik = require('fontnik');
const queue = require('d3-queue').queue;
const zlib = require('zlib');

/**
 * Make all metadata (codepoints) and SDF PBFs necessary for Mapbox GL fontstacks.
 * @param {Object} opts Options object with required `font` and `filetype` properties, and any optional parameters.
 * @param {Buffer} opts.font The font file as a `Buffer`.
 * @param {string} opts.filetype Type of font (e.g. `'.ttf'` or `'.otf'`).
 * @param {number} [opts.concurrency] Concurrency to use when processing font into PBFs. If `undefined`, concurrency is unbounded.
 * @param {function(err, result)} callback Callback takes arguments (`error`, `result`).
 *
 * **Callback `result`:**
 * * `{string}` `name` The name of this font (concatenated family_name + style_name).
 * * `{Array}` `stack` An array of `{name: filename, data: buffer}` objects with SDF PBFs covering points 0-65535.
 * * `{Object}` `metadata` An object where `data` is a stringified codepoints result.
 * * `{name: string, data: Buffer}` `original` An object containing the original font file (named `"original{.filetype}"`)
 *
 */
function makeGlyphs(opts, callback) {
  // undefined (unset concurrency) means unbounded concurrency in d3-queue
  const q = queue(opts.concurrency);

  if (!(opts.font instanceof Buffer)) throw new Error('opts.font must be a Buffer');
  if (typeof opts.filetype !== 'string') throw new Error('opts.filetype must be a String');
  if (!(callback instanceof Function)) throw new Error('Callback must be a Function');

  fontnik.load(opts.font, (err, faces) => {
    if (err) {
      return callback(err);
    }

    if (faces.length > 1) {
      return callback(new Error('Multiple faces in a font are not yet supported.'));
    }

    const metadata = faces[0];

    for (let i = 0; i < 65536; (i = i + 256)) {
      q.defer(writeGlyphs, opts.font, i, Math.min(i + 255, 65535));
    }

    q.awaitAll((err, results) => {
      if (err) return callback(err);
      return callback(null, {
        name: metadata.style_name ? [metadata.family_name, metadata.style_name].join(' ') : metadata.family_name,
        stack: results,
        metadata: Object.keys(metadata).reduce((prev, key) => {
          if (key !== 'points') {
            prev[key] = metadata[key];
          }

          return prev;
        }, {}),
        codepoints: metadata.points,
        original: {
          name: 'original' + opts.filetype,
          data: opts.font
        }
      });
    });
  });
}

function writeGlyphs(font, start, end, done) {
  fontnik.range({ font: font, start: start, end: end }, (err, data) => {
    if (err) {
      return done(err);
    }
    const name = [start, '-', end, '.pbf'].join('');
    zlib.gzip(data, (err, res) => {
      if (err) {
        return done(err);
      }
      const result = {
        name: name,
        data: res
      };
      return done(null, result);
    });
  });
}

module.exports = {
  makeGlyphs: makeGlyphs,
  writeGlyphs: writeGlyphs // exported only for testing
};
